flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/BooleanToNumericCastRule.java

```diff
@@ -0,0 +1,102 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.functions.casting;
+
+import org.apache.flink.table.types.logical.DecimalType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeFamily;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+
+import static org.apache.flink.table.planner.codegen.CodeGenUtils.primitiveLiteralForType;
+import static org.apache.flink.table.planner.codegen.calls.BuiltInMethods.DECIMAL_ZERO;
+import static org.apache.flink.table.planner.codegen.calls.BuiltInMethods.INTEGRAL_TO_DECIMAL;
+import static org.apache.flink.table.planner.functions.casting.CastRuleUtils.staticCall;
+import static org.apache.flink.table.planner.functions.casting.CastRuleUtils.ternaryOperator;
+
+/** {@link LogicalTypeRoot#BOOLEAN} to {@link LogicalTypeFamily#NUMERIC} conversions. */
+public class BooleanToNumericCastRule
+        extends AbstractExpressionCodeGeneratorCastRule<Boolean, Number> {
+
+    static final BooleanToNumericCastRule INSTANCE = new BooleanToNumericCastRule();
+
+    private BooleanToNumericCastRule() {
+        super(
+                CastRulePredicate.builder()
+                        .input(LogicalTypeRoot.BOOLEAN)
+                        .target(LogicalTypeFamily.NUMERIC)
+                        .build());
+    }
+
+    @Override
+    public String generateExpression(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        return ternaryOperator(
+                inputTerm, trueValue(targetLogicalType), falseValue(targetLogicalType));
+    }
+
+    private String trueValue(LogicalType target) {
+        switch (target.getTypeRoot()) {
+            case DECIMAL:
+                DecimalType decimalType = (DecimalType) target;
+                return staticCall(
+                        INTEGRAL_TO_DECIMAL(),
+                        1,
+                        decimalType.getPrecision(),
+                        decimalType.getScale());
+            case TINYINT:
+                return primitiveLiteralForType((byte) 1);
+            case SMALLINT:
+                return primitiveLiteralForType((short) 1);
+            case INTEGER:
+                return primitiveLiteralForType(1);
+            case BIGINT:
+                return primitiveLiteralForType(1L);
+            case FLOAT:
+                return primitiveLiteralForType(1f);
+            case DOUBLE:
+                return primitiveLiteralForType(1d);
+        }
+        throw new IllegalArgumentException("This is a bug. Please file an issue.");
```

Comment:
nit, don't have to change: It's more of personal taste to use default statement to throw the exception, also below, but it's up to you.